test/boost/reader_concurrency_semaphore_test.cc

```diff
@@ -1077,3 +1084,352 @@ SEASTAR_THREAD_TEST_CASE(test_reader_concurrency_semaphore_set_resources) {
+            return seastar::async([&, i] {
+                while (num_sstables != target_num_sstables) {
+                    ++num_sstables;
+                    auto sst = sst_man.make_sstable(s, sstables_dir.path().string(), sstables::generation_type{num_sstables}, sst_man.get_highest_supported_format(), sstables::sstable_format_types::big);
+                    auto writer_cfg = sst_man.configure_writer("test");
+                    sst->write_components(
+                        make_flat_mutation_reader_from_mutations_v2(s, semaphore.make_tracking_only_permit(s.get(), "test", db::no_timeout), mut, s->full_slice()),
+                        1,
+                        s,
+                        writer_cfg,
+                        encoding_stats{}).get();
+                    sst->load().get();
+                    tbl.add_sstable_and_update_cache(std::move(sst)).get();
+                }
+            });
+        }).get();
+
+        auto read_id = env.prepare("SELECT value FROM ks.tbl WHERE pk = ? AND ck = ?").get0();
+
+        // We first check that the test params are not too strict and a single
+        // read can finish successfully.
+        {
+            auto msg = env.execute_prepared(read_id, {cql3::raw_value::make_value(serialized(0)), cql3::raw_value::make_value(serialized(0))}).get();
+            assert_that(msg).is_rows().with_rows_ignore_order({ {serialized(value)} });
+        }
+
+        uint64_t successful_reads = 0;
+        uint64_t failed_reads = 0;
+
+        parallel_for_each(boost::irange(0, num_reads), [&] (int i) {
+            return env.execute_prepared(read_id, {cql3::raw_value::make_value(serialized(0)), cql3::raw_value::make_value(serialized(0))}).then_wrapped([&] (future<shared_ptr<cql_transport::messages::result_message>> fut) {
+                if (fut.failed()) {
+                    // We expect failed, OOM-killed reads here.
+                    // No way to verify why they failed so we swallow all failures.
+                    fut.ignore_ready_future();
+                    ++failed_reads;
+                    return;
+                }
+                assert_that(fut.get()).is_rows().with_rows_ignore_order({ {serialized(value)} });
+                ++successful_reads;
+            });
+        }).get();
+
+        testlog.info("total reads: {} ({} successful, {} failed)", num_reads, successful_reads, failed_reads);
+        dump_stats(semaphore, log_level::info, false);
+
+        // There should be both successful and failed reads.
+        // If there is only one or the other, the test is not testing anything.
+        // We also check that the memory limiting mechanism of the semaphore was engaged.
+        // The test is meaningless without it.
+        // In the slow debug builds we never reach the kill limit for some reason.
+#ifndef DEBUG
+        BOOST_REQUIRE_GE(failed_reads, 1);
+#endif
+        BOOST_REQUIRE_GE(successful_reads, 1);
```

Comment:
Ok, pre-patch we may not have a successful read, but on the other hand, we also may.

Reply:
It took a lot of tuning cycles to get both failed and successful reads. The current parameters seem to work reliably. I hope we have enough wiggle room here, we really don't need more flaky tests in CI.
I'll experiment with the params a bit to see what are the lower and upper bounds (if I can find them).